CAP/gap/TerminalCategory.gi

```diff
@@ -4,55 +4,414 @@
+    
+    Append( list_of_operations_to_install,
+            [ "Colift",
+              "Lift",
+              "IsColiftable",
+              "IsLiftable",
+               ] );
+    
+    for r in
+      [ "RIGID_SYMMETRIC_CLOSED_MONOIDAL_CATEGORIES_METHOD_NAME_RECORD",
+        "RIGID_SYMMETRIC_COCLOSED_MONOIDAL_CATEGORIES_METHOD_NAME_RECORD",
+        "CATEGORY_CONSTRUCTOR_METHOD_NAME_RECORD",
+        "FREYD_CATEGORIES_METHOD_NAME_RECORD",
+        "BOOLEAN_ALGEBRA_METHOD_NAME_RECORD",
+        ] do
+        
+        if IsBoundGlobal( r ) then
+            Append( list_of_operations_to_install, RecNames( ValueGlobal( r ) ) );
+        fi;
+        
+    od;
+    
+    list_of_operations_to_install := Set( list_of_operations_to_install );
+    
+    for operation_name in ShallowCopy( list_of_operations_to_install ) do
+        info := CAP_INTERNAL_METHOD_NAME_RECORD.(operation_name);
+        
+        ## do not install universal morphisms but their
+        ## with-given-universal-object counterpart and the universal object
+        if info.return_type = "morphism" and IsList( info.with_given_without_given_name_pair ) and operation_name = info.with_given_without_given_name_pair[1] then
+            if not info.with_given_without_given_name_pair[2] in list_of_operations_to_install then
+                Add( list_of_operations_to_install, info.with_given_without_given_name_pair[2] );
+            fi;
+            if IsBound( CAP_INTERNAL_METHOD_NAME_RECORD.(info.with_given_without_given_name_pair[2]).with_given_object_name ) then
+                Remove( list_of_operations_to_install, Position( list_of_operations_to_install, operation_name ) );
+                with_given_object_name := CAP_INTERNAL_METHOD_NAME_RECORD.(info.with_given_without_given_name_pair[2]).with_given_object_name;
+                if not with_given_object_name in list_of_operations_to_install then
+                    Add( list_of_operations_to_install, with_given_object_name );
+                fi;
+            fi;
+        fi;
+        
+    od;
+    
+    list_of_operations_to_install := Set( list_of_operations_to_install );
+    
+    skip := [ "IsEqualForObjects",
+              "IsEqualForMorphisms",
+              "IsCongruentForMorphisms",
+              "LeftCartesianDistributivityExpandingWithGivenObjects", ## gets derived since IsDistributiveCartesianCategory := true
+              "LeftCocartesianCodistributivityFactoringWithGivenObjects", ## gets derived since IsCodistributiveCocartesianCategory := true
+              "MorphismToBidual",
+              "MorphismFromBidual",
+              "MorphismToCoBidual",
+              "MorphismFromCoBidual",
+              "MorphismToBidualWithGivenBidual",
+              "MorphismFromBidualWithGivenBidual",
+              "MorphismToCoBidualWithGivenCoBidual",
+              "MorphismFromCoBidualWithGivenCoBidual",
```

Comment:
I don't remember :)

Reply:
There is a comment above: "gets derived since IsCodistributiveCocartesianCategory := true"

Reply:
The comment only applies to the forth and fifth element of the list. And I checked: `MorphismToBidual` is not even in `list_of_operations_to_install`, so this does not seem clean.

If I see this correctly, the whole logic above does this:

1. Take the operations from CAP_INTERNAL_CONSTRUCTIVE_CATEGORIES_RECORD.
2. Add some hardcoded operations which do not belong to a doctrine (e.g. `Lift`).
3. For some hardcoded packages which do not implement proper doctrines: Just take everything.
4. Manually remove some of the operations added in 3. again to test some derivations.

The last three points do not scale -> packages should have a way for registering operations for the terminal category. For example, we could add a new entry `install_for_terminal_category` to the method record.

Reply:
>For example, we could add a new entry install_for_terminal_category to the method record.

I think one would need two entries, since the two terminal categories don't share all mathematical properties. But I'm totally with you, this is currently ugly.

Reply:
But does this affect the operations?

Even if yes: `IsTerminalCategory` is a categorical property, so we should simply mark those available for every terminal category and not care about a particular implementation.

Reply:
I think my main trouble is, that I don't understand how categorical properties should behave for equivalent categories. 

I would assume, that equivalent categories should share the same categorical properties. However, despite being equivalent, the two terminal categories do not share all categorical properties, for example `IsSkeletalCategory`. So if we were to mark operations in `Locales`, which are only supposed for skeletal categories (e.g. `IsHeytingAlgebra`), for the terminal category with a single object, these operations would also be installed for the terminal category with multiple objects? But this would be wrong?

Maybe the problem is, that the current implementation of the terminal category with multiple objects should not be `IsTerminalCategory`. But this would be stupid.

Or do you mean, we mark those operations available for every terminal category, except for operations requiring strictness and skeletality? But then the terminal category with a single object wouldn't "possess its full potential", which also sounds stupid to me.

I guess I'm confusing myself.  :)

Reply:
> I would assume, that equivalent categories should share the same categorical properties.

I agree that this is what one should expect of a "proper" categorical property like `IsAdditiveCategory`. If we want to distinguish them, what would we call properties like `IsSkeletalCategory`? (This question is aimed at everyone reading this :D ) Making this distinction would probably also solve the problem that we have to manually filter the categorical properties which we set for the terminal category with multiple objects: We could probably just take the "proper" categorical properties.

> So if we were to mark operations in Locales, which are only supposed for skeletal categories (e.g. IsHeytingAlgebra), for the terminal category with a single object, these operations would also be installed for the terminal category with multiple objects? But this would be wrong?

As just discussed, those operations also make sense in the non-skeletal case.

> But then the terminal category with a single object wouldn't "possess its full potential", which also sounds stupid to me.

If this is stupid or not depends on what we aim for, I would say :D The aim I have in mind is merging the two existing implementations (in a clean way!) without losing functionality. Since the existing terminal category (with single object) has very little functionality, this is not too hard. Any extension to its "full potential" could then be made afterwards :-)

Reply:
> > I would assume, that equivalent categories should share the same categorical properties.
> 
> I agree that this is what one should expect of a "proper" categorical property like `IsAdditiveCategory`. If we want to distinguish them, what would we call properties like `IsSkeletalCategory`? (This question is aimed at everyone reading this :D ) Making this distinction would probably also solve the problem that we have to manually filter the categorical properties which we set for the terminal category with multiple objects: We could probably just take the "proper" categorical properties.

An idea that came from @mohamed-barakat while converting categorical properties from `Locales` into doctrines is to introduce another record collecting proper categorical doctrines (or maybe just install a new filter `IsCategoricalDoctrine` for categorical properties that are actually doctrines). The terminal category with multiple objects could then install its operations based on doctrines, and as you say this could completely eliminate the problems with `Skeletal` and `Strict`. Additionally, we would not need to go into every method record to mark operations for categories which are `IsTerminalCategory` and the terminal category with a single object could still just install anything.

Reply:
I have removed the skipped operations and the various lifts for now. (I'm commenting here, so you know which ones).
Is this fine for you? At least until the method records are differently organized.

Reply:
I think we should still skip `IsEqualForObjects`, `IsEqualForMorphisms`, and `IsCongruentForMorphisms`, or are those handled correctly despite not being skipped here?

Reply:
They are installed explicitly after the call to CategoryConstructor:
https://github.com/homalg-project/CAP_project/blob/83084b2a72ad94d349a67ab56615100e854d425f/CAP/gap/TerminalCategory.gi#L100-L111
https://github.com/homalg-project/CAP_project/blob/83084b2a72ad94d349a67ab56615100e854d425f/CAP/gap/TerminalCategory.gi#L217-L234
https://github.com/homalg-project/CAP_project/blob/83084b2a72ad94d349a67ab56615100e854d425f/CAP/gap/TerminalCategory.gi#L356-L374
So, I think yes? This manual example also works:
```gap
gap> t := TerminalCategoryWithMultipleObjects();
TerminalCategoryWithMultipleObjects( )
gap> a := "a" / t;
<A zero object in TerminalCategoryWithMultipleObjects( )>
gap> b := "b" / t;
<A zero object in TerminalCategoryWithMultipleObjects( )>
gap> m := MorphismConstructor( a, "", b );
<A zero, isomorphism in TerminalCategoryWithMultipleObjects( )>
gap> n := MorphismConstructor( a, "g", b );
<A zero, isomorphism in TerminalCategoryWithMultipleObjects( )>
gap> IsCongruentForMorphisms(m,n);
true
gap> IsCongruentForMorphisms(m,m);
true
gap> IsEqualForMorphisms(m,n);
false
gap> IsEqualForMorphisms(m,m);
true
gap> IsEqualForObjects(a,b);
false
gap> IsEqualForObjects(a,a);
true

```

Reply:
As discussed on Friday, the correct operations overwrite the wrongly installed ones, but I do not want to rely on this behaviour. Let's remove `create_func_bool` completely and reintroduce things we need in the future in a clean way.

For future usage: I just learned about a method record flag which I did not know before: `is_reflected_by_faithful_functor`. This might be useful in this context.

Reply:
Completely Removing `create_func_bool` (or only using `default` as written in the documentation) gave me errors, so I interpreted this as "don't install boolean operations":
https://github.com/homalg-project/CAP_project/blob/24148d49089e2cd5b5b0e06fb80b10b4441d5c92/CAP/gap/TerminalCategory.gi#L62-L65
 Is this right?

For the record: the boolean operations missing after this are

- `IsAutomorphism`
- `IsEpimorphism`
- `IsInitial`
- `IsIsomorphism`
- `IsMonomorphism`
- `IsSplitEpimorphism`
- `IsSplitMonomorphism`
- `IsTerminal`

The `IsSplit*` -versions are enough to add. Would you like them in a different PR or also this one?

Reply:
> Completely Removing create_func_bool (or only using default as written in the documentation) gave me errors, so I interpreted this as "don't install boolean operations":

Correct! And after excluding all boolean operations you should be able to remove `create_func_bool` :-)

> The `IsSplit*`-versions are enough to add. Would you like them in a different PR or also this one?

You mean for the existing tests to pass? Then please add them here (with a comment that this should be generalized in the future) so we do not get any regressions.

Reply:
> > Completely Removing create_func_bool (or only using default as written in the documentation) gave me errors, so I interpreted this as "don't install boolean operations":
> 
> Correct! And after excluding all boolean operations you should be able to remove `create_func_bool` :-)

Thanks! I somehow didn't have this idea.

> > The `IsSplit*`-versions are enough to add. Would you like them in a different PR or also this one?
> 
> You mean for the existing tests to pass? Then please add them here (with a comment that this should be generalized in the future) so we do not get any regressions.

Ah, no, the tests run fine without it. :) I just meant adding the `IsSplit*`- versions derives all the boolean operations in the list above.

Reply:
> Ah, no, the tests run fine without it.

Then please do *not* add the properties to reduce the number of special cases. If anyone needs it later, we can discuss how to add it properly.

Reply:
I don't understand: Why can't we simply manually `AddIsSplitMonomorphism` and `AddIsSplitEpimorphisms` to return `true`? I would be strongly in favor of adding them.

Reply:
I thought the result of our decision was that we do not add special cases as long as the tests pass, and re-add them later in a more generic way if the need arises.

Besides, I just checked that `IsSplitMonomorphism` and `IsSplitEpimorphism` never were part of `list_of_operations_to_install`. They were previously derived from `IsLiftable` and `IsColiftable`. All the other properties listed above are actually computable. I don't know why in @TKuh's tests they were not available. Maybe due to properties not being set correctly, which was fixed in https://github.com/homalg-project/CAP_project/blob/274ba63a3b8bac4ae7751bc34ed2e974544b1537/CAP/gap/TerminalCategory.gi#L83

Reply:
Thank you for your analysis. Maybe for some operations, we do not need a high level of genericity. For those, I favor a minimal pragmatic approach that can be amended once we have an idea for a generic way. I suggest manually adding `IsLiftable`, `IsColiftable`, `Lift`, `Colift`.

Reply:
I think for those operations the generic solution is easy: Simply add categorical properties `IsCategoryWithLifts` (or `IsCategoryWithDecidableLifts` or something else?) and add those to `CAP_INTERNAL_CONSTRUCTIVE_CATEGORIES_RECORD`.

Reply:
This is indeed what I am doing now :)